### PR TITLE
Table block

### PR DESCRIPTION
**Pull request by seanchoi-dev**
New block: Table
- Design: https://www.figma.com/file/tZXvTXkgnUe1RSzVAsZyW5/Tables-Design-Sprint?type=design&node-id=405-792

Resolves: [MWPW-129568](https://jira.corp.adobe.com/browse/MWPW-129568)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/docs/library/blocks/table?martech=off
- After: https://tableblock--milo--adobecom.hlx.page/docs/library/blocks/table?martech=off

Documentation:
https://github.com/adobecom/milo/tree/tableBlock/libs/blocks/table#readme

### CODE REVIEW

**Comment by seanchoi-dev** [comment]
Added table.test.js with 100% coverage.
image

File: libs/blocks/table/table.js [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+  const tableHighlightLoadedEvent = new Event('milo:table:highlight:loaded');
+  window.dispatchEvent(tableHighlightLoadedEvent);
+}
+
+function handleExpand(e) {
+  const sectionHead = e.closest('.row');
+  let nextElement = sectionHead.nextElementSibling;
+  const expanded = e.getAttribute('aria-expanded') === 'false';
+  e.setAttribute('aria-expanded', expanded.toString());
+  while (nextElement && !nextElement.classList.contains('devider')) {
```
Comment (chrischrischris) [inline libs/blocks/table/table.js:93]: typo - should be `divider`

I think you should use consts for the classes that are used multiple times

File: libs/blocks/table-metadata/table-metadata.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      element.style.backgroundColor = color;
+      if (columnType === 'highlight') {
+        element.style.borderColor = color;
+      }
+    });
+  } else {
+    bgColors.forEach((color) => {
+      const [bgColorIndex, bgColorValue] = color;
+      const col = table.querySelector(`.col-${bgColorIndex}.col-${columnType}`);
+      if (col && col.innerText) {
```
Comment (chrischrischris) [inline libs/blocks/table-metadata/table-metadata.js:40]: nit: All of the `obj && obj.property` checks can be changed to `obj?.property`

File: libs/blocks/table/table.js [hunk 3]
```diff
@@ -0,0 +1,430 @@
+/* eslint-disable no-plusplus */
+import { createTag } from '../../utils/utils.js';
+import { decorateButtons } from '../../utils/decorate.js';
+
+function handleHeading(headingCols) {
+  headingCols.forEach((col) => {
+    if (col.innerHTML) {
```
Comment (chrischrischris) [inline libs/blocks/table/table.js:7]: nit: early return

File: libs/blocks/table/table.js [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+    if (highlightRow) highlightRow.style.top = `${gnavHeight}px`;
+    headingRow.style.top = `${gnavHeight + (highlightRow ? highlightRow.offsetHeight : 0)}px`;
+  });
+  observer.observe(intercept);
+}
+
+function applyStylesBasedOnScreenSize(table, originTable) {
+  const isMerch = table.classList.contains('merch');
+  const desktopSize = 900;
+  const mobileSize = 768;
```
Comment (chrischrischris) [inline libs/blocks/table/table.js:279]: nit
```suggestion
  const DESKTOP_SIZE = 900;
  const MOBILE_SIZE = 768;
```
could also be `SIZE_DESKTOP` etc

**Review by chrischrischris — COMMENTED** [review]
Only have time for a quick initial review.

Some of the funcs are really long and could use some refactoring to break them up into multiple smaller funcs, especially the large if/else blocks.

Any text values that are used multiple times should be converted to top level consts.

**Comment by meganthecoder** [comment]
Have you done an accessibility review on this? When we did the comparison table for bacom, we got this advise from Taryn:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/table_role

You can use <div>, but you have to build in the accessibility manually, whereas it comes for free using <table>. Using native elements is always best whenever possible.

https://www.w3.org/WAI/tutorials/tables/

Relying on visual cues alone is not sufficient to create an accessible table.

**Comment by auniverseaway** [comment]
@seanchoi-dev @Blainegunn Are you planning addressing @meganthecoder's comments now or are you filing tickets for later? Either way, should this be considered for a merge or is this still in progress?

**Comment by seanchoi-dev** [comment]
> 

@auniverseaway @meganthecoder  I've added some Accessibility work.

File: libs/blocks/table/table.js [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+
+function handleMouseOut(cols) {
+  cols.forEach((e) => {
+    e.classList.remove('hover');
+    e.classList.remove('no-top-border');
+    e.classList.remove('hover-border-bottom');
+  });
+}
+
+function handleMouseOver(cols, table, colNum, isCollapseTable, lastSectionHead, lastExpandIcon) {
```
Comment (meganthecoder) [inline libs/blocks/table/table.js:226]: This seems like a lot of params. Could it be paired down or would a destructured object be helpful?

File: libs/blocks/table/table.js [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+  const headingRow = table.querySelector('.row-heading');
+  const colClass = `col-${colNum}`;
+  const isLastRowCollapsed = lastExpandIcon?.getAttribute('aria-expanded') === 'false';
+
+  cols.forEach((e) => {
+    if (e.classList.contains('col-highlight') && e.innerText) {
+      const matchingCols = Array.from(e.classList).filter(
+        (className) => className.startsWith(colClass),
+      );
+      matchingCols.forEach((className) => {
```
Comment (meganthecoder) [inline libs/blocks/table/table.js:238]: I'm a little confused about what's going on here. When would matchingCols every have more than one item?

File: libs/blocks/table/table.js [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+
+function handleScrollEffect(table, gnavHeight) {
+  const highlightRow = table.querySelector('.row-highlight');
+  const headingRow = table.querySelector('.row-heading');
+
+  if (highlightRow) {
+    highlightRow.style.top = `${gnavHeight}px`;
+    highlightRow.style.borderTop = '1px solid transparent';
+  } else {
+    headingRow.style.borderTop = '1px solid transparent';
```
Comment (meganthecoder) [inline libs/blocks/table/table.js:280]: Is there a reason this is added directly in the javascript? For maintainability, it'd be better to use a class and move these to the stylesheet

File: libs/blocks/table/table.js [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+  }
+  headingRow.style.top = `${gnavHeight + (highlightRow ? highlightRow.offsetHeight : 0)}px`;
+
+  const intercept = table.querySelector('.intercept') || document.createElement('div');
+  intercept.className = 'intercept';
+  intercept.setAttribute('data-observer-intercept', '');
```
Comment (meganthecoder) [inline libs/blocks/table/table.js:286]: createTag?

**Review by meganthecoder — COMMENTED** [review]
Lots of good stuff! I haven't finished my review, but here are my comments so far

**Comment by seanchoi-dev** [comment]
@draganatrajkovic Megan has made lots of comments on your code (Thanks! @meganthecoder ). I've resolved most of them but there are few more things that I want you to take a look and answer her reviews.

**Review by meganthecoder — COMMENTED** [review]
Nice! Lots of code! I have a couple high-level thoughts in addition to the code comments:
- It seems like there are a lot of repeated dom look-ups and loops. This is something that's been pointed out in scripts.js as being undesirable for performance reasons. I'm not sure if that's so much of a concern here (performance scores look decent), but I thought I'd call it out.
- The table metadata block is interesting. Blocks typically shouldn't affect other blocks, right? What was the reasoning behind creating a unique metadata block?

File: libs/blocks/table/table.js [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+}
+
+function handleSection(table) {
+  const isCollapseTable = table.classList.contains('collapse') && !table.classList.contains('merch');
+  const isHighlightTable = table.classList.contains('highlight');
+  const isMerch = table.classList.contains('merch');
+  const allRows = Array.from(table.getElementsByClassName('row'));
+  let defaultExpandRow = true;
+
+  allRows.forEach((row, i) => {
```
[the remark this answers is not in the document]
Comment (draganatrajkovic) [reply libs/blocks/table/table.js]: @meganthecoder can you please check my fix?

File: libs/blocks/table/table.js [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+        if (e.key === 'Enter' || e.key === ' ') handleExpand(e.target);
+      });
+    });
+    handleHovering(tableEl);
+  };
+
+  const mobileRenderer = () => {
+    if (isMerch && table.querySelectorAll('.row-heading .col').length > 2) {
+      table.querySelectorAll('.col:not(.col-1, .col-2)').forEach((col) => col.remove());
+    } else if (table.querySelectorAll('.row-heading .col').length > 3) {
```
[the remark this answers is not in the document]
Comment (draganatrajkovic) [reply libs/blocks/table/table.js]: @meganthecoder this is resloved

File: libs/blocks/table/table.js [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+      } else {
+        table.querySelectorAll(`.col:not(.col-1, .col-${filters[0] + 1}, .col-${filters[1] + 1}), .col.no-borders`).forEach((col) => col.remove());
+      }
+      if (filters[0] > filters[1]) {
+        if (isMerch) {
+          table.querySelectorAll('.row').forEach((row) => {
+            row.querySelector('.col:not(.section-row-title)').style.order = 1;
+          });
+        } else {
+          table.querySelectorAll('.row').forEach((row) => {
```
[the remark this answers is not in the document]
Comment (draganatrajkovic) [reply libs/blocks/table/table.js]: @meganthecoder this is resloved

File: libs/blocks/table/table.js [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+      highlights.forEach((highlight, index) => {
+        if (!highlight.innerHTML && !headings[index + 1]?.classList.contains('hidden')) {
+          table.querySelector('.row-heading').querySelectorAll(`.col-${index + 1}`).forEach((heading) => {
+            heading.classList.add('top-left-rounded', 'top-right-rounded');
+          });
+        }
+      });
+    };
+
+    // filter
```
[the remark this answers is not in the document]
Comment (draganatrajkovic) [reply libs/blocks/table/table.js]: @meganthecoder this is resloved

File: libs/blocks/table/table.js [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+      const filter2 = createTag('div', { class: 'filter-wrapper' });
+      const colSelect0 = createTag('select', { class: 'filter' });
+      const headings = originTable.querySelectorAll('.col-heading');
+      headings.forEach((heading, index) => {
+        const title = heading.querySelector('.heading-title');
+        if (!title || (!isMerch && title.closest('.col-1'))) return;
+
+        const option = createTag('option');
+        option.value = index;
+        option.innerHTML = title.innerText;
```
[the remark this answers is not in the document]
Comment (draganatrajkovic) [reply libs/blocks/table/table.js]: @meganthecoder this is resloved

File: libs/blocks/table/table.js [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+    const percentage = 100 / colsForTablet;
+    const templateColumnsValue = `repeat(auto-fit, ${percentage}%)`;
+    sectionRow.forEach((row) => {
+      if (isMerch) {
+        row.style.gridTemplateColumns = '';
+      } else if (deviceBySize === 'TABLET') {
+        row.style.gridTemplateColumns = templateColumnsValue;
+      } else {
+        row.style.gridTemplateColumns = '';
+      }
```
[the remark this answers is not in the document]
Comment (draganatrajkovic) [reply libs/blocks/table/table.js]: @meganthecoder this is resolved, can you please check?

**Review by Blainegunn — APPROVED** [review]

**Comment by seanchoi-dev** [comment]
@meganthecoder Thanks for the reviews. @draganatrajkovic and I made few commits and they are all resolved.

For your question about table metadata,, it was needed to handle text & bg color control for authors.
We could not found a better way than that since we are trying to make this block as much as intuitive like table (doc) to table (web page). And of course, there are defaults if there is no table metadata.

File: test/blocks/table/table.test.js [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  });
+
+  describe('standard table', () => {
+    const table = document.querySelector('.table');
+
+    it('row-heading', () => {
+      expect(table.querySelector('.row-heading')).to.exist;
+    });
+
+    it('expand icon event by mouse click', () => {
```
Comment (meganthecoder) [inline test/blocks/table/table.test.js:22]: Nit: Tests should start with a verb in the correct tense and make sense as a sentence
```suggestion
    it('adds the row heading', () => {
      expect(table.querySelector('.row-heading')).to.exist;
    });

    it('expands on icon mouse click', () => {
```

**Pull request merged**